homeassistant/components/bluetooth/models.py

```diff
@@ -0,0 +1,142 @@
+
+MAX_HISTORY_SIZE: Final = 512
+
+
+def _dispatch_callback(
+    callback: AdvertisementDataCallback,
+    filters: dict[str, set[str]],
+    device: BLEDevice,
+    advertisement_data: AdvertisementData,
+) -> None:
+    """Dispatch the callback."""
+    if not callback:
+        # Callback destroyed right before being called, ignore
+        return
+
+    if (uuids := filters.get(FILTER_UUIDS)) and not uuids.intersection(
+        advertisement_data.service_uuids
+    ):
+        return
+
+    try:
+        callback(device, advertisement_data)
+    except Exception:  # pylint: disable=broad-except
+        _LOGGER.exception("Error in callback: %s", callback)
+
+
+class HaBleakScanner(BleakScanner):  # type: ignore[misc]
+    """BleakScanner that cannot be stopped."""
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        """Initialize the BleakScanner."""
+        self._callbacks: list[
+            tuple[AdvertisementDataCallback, dict[str, set[str]]]
+        ] = []
+        self._history: LRU = LRU(MAX_HISTORY_SIZE)
+        super().__init__(*args, **kwargs)
+
+    @hass_callback
+    def async_register_callback(
+        self, callback: AdvertisementDataCallback, filters: dict[str, set[str]]
+    ) -> CALLBACK_TYPE:
+        """Register a callback."""
+        callback_entry = (callback, filters)
+        self._callbacks.append(callback_entry)
+
+        @hass_callback
+        def _remove_callback() -> None:
+            self._callbacks.remove(callback_entry)
+
+        # Replay the history since otherwise we miss devices
+        # that were already discovered before the callback was registered
+        # or we are in passive mode
+        for device, advertisement_data in self._history.values():
+            _dispatch_callback(callback, filters, device, advertisement_data)
+
+        return _remove_callback
+
+    def async_callback_dispatcher(
+        self, device: BLEDevice, advertisement_data: AdvertisementData
+    ) -> None:
+        """Dispatch the callback.
+
+        Here we get the actual callback from bleak and dispatch
+        it to all the wrapped HaBleakScannerWrapper classes
+        """
+        self._history[device.address] = (device, advertisement_data)
+        for callback_filters in self._callbacks:
+            _dispatch_callback(*callback_filters, device, advertisement_data)
+
+
+class HaBleakScannerWrapper(BleakScanner):  # type: ignore[misc]
+    """A wrapper that uses the single instance."""
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        """Initialize the BleakScanner."""
+        self._detection_cancel: CALLBACK_TYPE | None = None
+        self._mapped_filters: dict[str, set[str]] = {}
+        if "filters" in kwargs:
+            self._mapped_filters = {k: set(v) for k, v in kwargs["filters"].items()}
+        if "service_uuids" in kwargs:
+            self._mapped_filters[FILTER_UUIDS] = set(kwargs["service_uuids"])
+        super().__init__(*args, **kwargs)
```

Comment:
How does this really work. Bleak does allow multiple active scanners with dbus connections. Here the scanner is inited but never started. So that scanner is never going to do anything?

Reply:
Ok i see what you do, you replace the base scanner. It likely should inherit from the bleak base class instead.

Reply:
https://github.com/home-assistant/core/pull/74773